### PR TITLE
build: improve script and travis config

### PR DESCRIPTION
**Pull request by lzutao**
* fix a diff failure on windows

See https://travis-ci.com/rust-lang/rust-clippy/jobs/245971932#L1625
for an example.

* use cmp instead of diff > /dev/null

* clone single branch instead of clone then checking out

* do not decrypt key if have no diff change

changelog: none

### CODE REVIEW

File: .github/deploy.sh [hunk 1]
```diff
@@ -27,35 +23,44 @@ mkdir out/master/
...
-        import os, json;\
-        print json.dumps([\
-            dir for dir in os.listdir(".")\
-            if not dir.startswith(".") and os.path.isdir(dir)\
-        ])' > versions.json
-)
+cp util/gh-pages/versions.html out/index.html
+pushd out
+
+cat <<-EOF | python - > versions.json
```
Comment (flip1995) [inline .github/deploy.sh]: IMO the `python -c` version is way easier to understand. With this I only can guess what it does. Does this change fix something?

I like the `pushd`/`popd` change though.

File: .github/deploy.sh [hunk 2]
```diff
@@ -64,19 +69,9 @@ ENCRYPTED_KEY=${!ENCRYPTED_KEY_VAR}
 ENCRYPTED_IV=${!ENCRYPTED_IV_VAR}
 openssl aes-256-cbc -K "$ENCRYPTED_KEY" -iv "$ENCRYPTED_IV" -in .github/deploy_key.enc -out .github/deploy_key -d
 chmod 600 .github/deploy_key
-eval $(ssh-agent -s)
+eval "$(ssh-agent -s)"
```
Comment (flip1995) [inline .github/deploy.sh]: (Why) Is this necessary?

File: ci/base-tests.sh [hunk 3]
```diff
@@ -1,16 +1,18 @@
...
+if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+  remark -f ./*.md -f doc/*.md > /dev/null
 fi
 # build clippy in debug mode and run tests
 cargo build --features "debugging deny-warnings"
 cargo test --features "debugging deny-warnings"
 # for faster build, share target dir between subcrates
-export CARGO_TARGET_DIR=`pwd`/target/
+CARGO_TARGET_DIR=$(pwd)/target/
+export CARGO_TARGET_DIR
```
Comment (flip1995) [inline ci/base-tests.sh]: Why?

File: tests/ui-toml/update-all-references.sh [hunk 4]
```diff
@@ -13,6 +13,6 @@ if [[ "$1" == "--help" || "$1" == "-h" ]]; then
 fi
 
 BUILD_DIR=$PWD/target/debug/test_build_base
-MY_DIR=$(dirname $0)
-cd $MY_DIR
-find . -name '*.rs' | xargs ./update-references.sh $BUILD_DIR
+MY_DIR=$(dirname "$0")
+cd "$MY_DIR" || exit
+find . -name '*.rs' -exec ./update-references.sh "$BUILD_DIR" {} +
```
Comment (flip1995) [inline tests/ui-toml/update-all-references.sh]: ~~IIRC, `--exec` only exists on newer versions of find and won't work with older versions.~~ 

Well it was new in 2005, so nevermind, I guess 😄 https://stackoverflow.com/a/897043

**Comment by flip1995** [comment]
Oh and thanks for all the work on the build system lately! ❤️

File: .github/deploy.sh (hunk 2 above)
Comment (lzutao) [reply .github/deploy.sh]: Not a functional change, but let's compare the result:
```console
% echo $(ssh-agent -s )
SSH_AUTH_SOCK=/tmp/ssh-uekMnQ1X5cAJ/agent.30019; export SSH_AUTH_SOCK; SSH_AGENT_PID=30020; export SSH_AGENT_PID; echo Agent pid 30020;
% echo "$(ssh-agent -s )"
SSH_AUTH_SOCK=/tmp/ssh-2LsmK7aT6yYE/agent.30210; export SSH_AUTH_SOCK;
SSH_AGENT_PID=30211; export SSH_AGENT_PID;
echo Agent pid 30211;
```

File: ci/base-tests.sh (hunk 3 above)
Comment (lzutao) [reply ci/base-tests.sh]: It is to avoid masking return values of pwd.

File: .github/deploy.sh (hunk 1 above)
Comment (lzutao) [reply .github/deploy.sh]: Look at it again. I think we could > out/version.json directly, no need for cd and pushd.

Comment (lzutao) [reply .github/deploy.sh]: Also I like this version better because you could write python script in bash as in a python file.

Comment (flip1995) [reply .github/deploy.sh]: Ok, let's do the `<<-EOF` version with `out/version.json`.

File: .github/deploy.sh [hunk 5]
```diff
@@ -27,35 +23,41 @@ mkdir out/master/
...
-        print json.dumps([\
-            dir for dir in os.listdir(".")\
-            if not dir.startswith(".") and os.path.isdir(dir)\
-        ])' > versions.json
-)
+cp util/gh-pages/versions.html out/index.html
+cat <<-EOF | python - > out/versions.json
+import os, json
+print json.dumps([
+    dir for dir in os.listdir(".") if not dir.startswith(".") and os.path.isdir(dir)
```
Comment (flip1995) [inline .github/deploy.sh]: Oh wait, I think the `cd out` was necessary for this, since `listdir(".")` now points to `out/..`

Comment (lzutao) [reply .github/deploy.sh]: Yes, that's true.

File: ci/base-tests.sh [hunk 6]
```diff
@@ -29,26 +31,27 @@ export CARGO_TARGET_DIR=`pwd`/target/
...
   sysroot=$(SYSROOT=$desired_sysroot ./target/debug/clippy-driver --print sysroot)
-  test $sysroot = $desired_sysroot
+  test "$sysroot" = $desired_sysroot
 
   # Make sure this isn't set - clippy-driver should cope without it
   unset CARGO_MANIFEST_DIR
 
   # Run a lint and make sure it produces the expected output. It's also expected to exit with code 1
-  # XXX How to match the clippy invocation in compile-test.rs?
-  ! ./target/debug/clippy-driver -Dwarnings -Aunused -Zui-testing --emit metadata --crate-type bin tests/ui/cstring.rs 2> cstring.stderr
```
Comment (lzutao) [inline ci/base-tests.sh]: I don't understand what this command does. Could someone explain it to me?

File: ci/integration-tests.sh [hunk 7]
```diff
@@ -1,24 +1,28 @@
...
+  --cap-lints warn -W clippy::pedantic -W clippy::nursery \
+  > clippy_output 2>&1
 
-case ${INTEGRATION} in
-  *)
-    check
-    ;;
-esac
+cat clippy_output
+if grep -q "internal compiler error\|query stack during panic\|E0463" clippy_output; then
```
Comment (lzutao) [inline ci/integration-tests.sh]: Similar, do this transition correct?

File: .github/deploy.sh (hunk 5 above)
Comment (lzutao) [reply .github/deploy.sh]: Reverted to the old behaviour.

File: ci/integration-tests.sh (hunk 7 above)
Comment (flip1995) [reply ci/integration-tests.sh]: I think so. I tested it locally and it seemed to work.

File: ci/base-tests.sh (hunk 6 above)
Comment (flip1995) [reply ci/base-tests.sh]: This whole block checks if `clippy-driver` can be run without cargo. 

This command is just a test, if the output of `clippy-driver` is the same as the output of the ui-test. Why `cstring.rs` was chosen for this test, I have no idea. 

I guess all these flags are there to simulate the behavior of `compiletest-rs`.

Comment (phansch) [reply ci/base-tests.sh]: The PR that added this has some more context. Maybe include the PR link in the comment?

File: tests/ui/update-references.sh [hunk 8]
```diff
@@ -26,21 +26,19 @@ while [[ "$1" != "" ]]; do
     STDOUT_NAME="${1/%.rs/.stdout}"
     FIXED_NAME="${1/%.rs/.fixed}"
     shift
-    if [ -f $BUILD_DIR/$STDOUT_NAME ] && \
-           ! (diff $BUILD_DIR/$STDOUT_NAME $MYDIR/$STDOUT_NAME >& /dev/null); then
-        echo updating $MYDIR/$STDOUT_NAME
-        cp $BUILD_DIR/$STDOUT_NAME $MYDIR/$STDOUT_NAME
+    if [[ -f "$BUILD_DIR"/"$STDOUT_NAME" ]] && \
+           ! (cmp -s -- "$BUILD_DIR"/"$STDOUT_NAME" "$MYDIR"/"$STDOUT_NAME"); then
```
Comment (phansch) [inline tests/ui/update-references.sh]: Is there a reason to choose `cmp -s` over `diff -q`?

Comment (flip1995) [reply tests/ui/update-references.sh]: According to stackexchange, no: https://unix.stackexchange.com/a/153612 (at least performance wise)

Comment (lzutao) [reply tests/ui/update-references.sh]: `cmp -s` is completely quiet, unlike `diff -q` that print a line to stdin confirm that files are differ.

**Comment by lzutao** [comment]
Ready to review!
I sorted out the test failure in cargo-clippy on macOS systems.
I will remove those WIP commits after this PR is approved or on demand.

File: .travis.yml [hunk 9]
```diff
@@ -64,82 +58,82 @@ matrix:
...
-    fi
+    SYSROOT=$(rustc --print sysroot)
+    case "$TRAVIS_OS_NAME" in
+      windows ) export PATH="${SYSROOT}/bin:${PATH}" ;;
+      linux ) export LD_LIBRARY_PATH="${SYSROOT}/lib" ;;
+      osx )
+        # See <https://github.com/nteract/nteract/issues/1523#issuecomment-301623519>
+        sudo mkdir -p /usr/local/lib
+        sudo find "$SYSROOT/lib" -maxdepth 1 -name '*.dylib' -exec ln -s {} /usr/local/lib \;
+        ;;
```
Comment (flip1995) [inline .travis.yml]: This wasn't necessary before. What changed?

File: ci/base-tests.sh [hunk 10]
```diff
@@ -29,26 +33,27 @@ export CARGO_TARGET_DIR=`pwd`/target/
...
-  diff <(sed -e 's,tests/ui,$DIR,' -e '/= help/d' cstring.stderr) tests/ui/cstring.stderr
+  # FIXME: How to match the clippy invocation in compile-test.rs?
+  ./target/debug/clippy-driver -Dwarnings -Aunused -Zui-testing --emit metadata --crate-type bin tests/ui/cstring.rs 2> cstring.stderr || true
```
Comment (flip1995) [inline ci/base-tests.sh]: I think the `! _` is the better solution for a command that _should_ fail.

Comment (lzutao) [reply ci/base-tests.sh]:
```suggestion
  test ! ./target/debug/clippy-driver -Dwarnings -Aunused -Zui-testing --emit metadata --crate-type bin tests/ui/cstring.rs 2> cstring.stderr
```
would be better I think?

File: .travis.yml (hunk 9 above)
Comment (lzutao) [reply .travis.yml]: Honestly, I don't know. I tried to set LD_LIBRARY_PATH, LIBRARY_PATH, DYLD_LIBRARY_PATH. All of
that still made this PR failed.

**Comment by lzutao** [comment]
Build succeed. I wanna more eyes and thoughts on https://github.com/rust-lang/rust-clippy/pull/4675#discussion_r336869907.

**Comment by lzutao** [comment]
Note that in the first try, the cache will be invalidated because of removing BASE_TESTS env.

**Comment by phansch** [comment]
Changes look good to me. I can't say anything about the dylib stuff, but it seems to work fine :man_shrugging: 

Thanks for all your build improvements!

@bors r+

**Pull request merged**